### PR TITLE
Fix segfault for malformed requests

### PR DESCRIPTION
**Pull request by GordonMcKinney**
I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

This PR fixes issue #2001 

Unit test results:
```
# cd /home/openresty-*/bundle/ngx_lua-*
# prove t/104-req-raw-header.t
t/104-req-raw-header.t .. ok       
All tests successful.
Files=1, Tests=234,  6 wallclock secs ( 0.06 usr  0.01 sys +  0.67 cusr  0.30 csys =  1.04 CPU)
Result: PASS
```

### CODE REVIEW

File: src/ngx_http_lua_headers.c [hunk 1]
```diff
@@ -167,7 +167,7 @@ ngx_http_lua_ngx_req_raw_header(lua_State *L)
     size = 0;
     b = c->buffer;
 
-    if (mr->request_line.data[mr->request_line.len] == CR) {
+    if (mr->request_line.len>0 && mr->request_line.data[mr->request_line.len] == CR) {
```
Comment (zhuizhuhaomeng) [inline src/ngx_http_lua_headers.c:170]:
```suggestion
    if (mr->request_line.len == 0) {
        return 0;
    }

    if (mr->request_line.data[mr->request_line.len] == CR) {
```

Comment (GordonMcKinney) [reply src/ngx_http_lua_headers.c:170]: I'll run some tests with this proposal.

Comment (GordonMcKinney) [reply src/ngx_http_lua_headers.c:170]: Running this code produces the following error:

```
2022/01/31 15:53:18 [error] 9#0: *1 failed to run log_by_lua*: log_by_lua(nginx.conf:62):2: bad argument #1 to 'len' (string expected, got nil)
stack traceback:
        [builtin:len]: in function 'len'
        log_by_lua(nginx.conf:62):2: in main chunk while logging request, client: 172.17.0.1, server: localhost
```

I am unable to push new commits to this PR without a lengthy legal, security, and manager approval. I had to ask for an emergency escalation to get this PR logged in just four days. I expect future commits to take 1-2 weeks for a complete review.

I am happy to assist in testing any further commit proposals you may have.

**Comment by GordonMcKinney** [comment]
Commit 3c9bcd7 test results:
```
# prove t/104-req-raw-header.t
t/104-req-raw-header.t .. ok       
All tests successful.
Files=1, Tests=234,  5 wallclock secs ( 0.06 usr  0.00 sys +  0.67 cusr  0.38 csys =  1.11 CPU)
Result: PASS
```

**Review by zhuizhuhaomeng — APPROVED** [review]

**Comment by zhuizhuhaomeng** [comment]
merged

**Comment by GordonMcKinney** [comment]
@zhuizhuhaomeng The PR indicates it's closed with unmerged commits.

**Comment by zhuizhuhaomeng** [comment]
I merge it with small tweaks and merged through cli rather than github web.
you can see commit 2615bafa34f

**Comment by GordonMcKinney** [comment]
@zhuizhuhaomeng I downloaded https://github.com/openresty/lua-nginx-module/releases/tag/v0.10.21 and could not find test 35 in t/104-req-raw-header.t. I also looked at src/ngx_http_lua_headers.c and line 170 -175 are unchanged.

Is this merge in a different release train?

**Comment by bjoe2k4** [comment]
@GordonMcKinney It's just in master, but not in any tagged release yet.

**Comment by GordonMcKinney** [comment]
@bjoe2k4 Thanks for the quick response. Is there an ETA for this fix being released? We have a downstream dependency (Kong) that must consume the fix to enable us to revert our production workaround.

**Comment by GordonMcKinney** [comment]
@bjoe2k4 bump. Kong just consumed 0.10.21. I'd love to have the segfault fix in a tagged release so that Kong can use it in the future. tia :-)

**Comment by bjoe2k4** [comment]
@GordonMcKinney I'm neither affiliated with this module nor do i have the necessary priviliges. I was just stating the obvious. Just don't use tags and use master. Or fork and create a tag.

**Comment by GordonMcKinney** [comment]
@zhuizhuhaomeng and @xiaocang our team identified this problem in production and provided a fix at the beginning of the year. @zhuizhuhaomeng kindly merged the fix into master in February.

We need this fix in the Kong code base and they just consumed v0.10.21 which doesn't have this code. How can we get this fix into a tagged release so that Kong may consume it?

**Pull request closed without merging**